src/redux/books/books.js

```diff
@@ -1,47 +1,85 @@
-import { v4 as uuidv4 } from 'uuid';
-
 // Actions
+const GET_BOOKS = 'bookstore-react/books/GET_BOOKS';
 const ADD = 'bookstore-react/books/ADD';
 const REMOVE = 'bookstore-react/books/REMOVE';
 
+// key = kcM4ut7Kvp9QgSxmujh0
+const API_KEY = 'kcM4ut7Kvp9QgSxmujh0';
+
 // Reducer
-export default function booksReducer(
-  state = [
-    {
-      title: 'First Book',
-      author: 'First Author',
-      id: uuidv4(),
-    },
-    {
-      title: 'Second Book',
-      author: 'Second Author',
-      id: uuidv4(),
-    },
-    {
-      title: 'Third Book',
-      author: 'Third Author',
-      id: uuidv4(),
-    },
-  ],
-  action = {}
-) {
+export default function booksReducer(state = [], action = {}) {
   switch (action.type) {
+    case GET_BOOKS:
+      return action.payload;
     case ADD:
       return [...state, action.book];
     case REMOVE:
-      return state.filter((book) => book.id !== action.book.id);
+      return state.filter((book) => book.item_id !== action.item_id);
     default:
       return state;
   }
 }
 
 // Action Creators
-export const addBook = (book) => ({
-  type: ADD,
-  book,
-});
+export const getBooks = () => async (dispatch) => {
+  const booksFetch = await fetch(
+    `https://us-central1-bookstore-api-e63c8.cloudfunctions.net/bookstoreApi/apps/${API_KEY}/books`,
+    {
+      method: 'GET',
+      headers: {
+        Accept: 'application/json',
+      },
+    },
+  ).then((data) => data.json());
+  const books = [];
+  Object.keys(booksFetch).forEach((e) => {
+    books.push({ ...booksFetch[e][0], item_id: e });
+  });
+  dispatch({
+    type: GET_BOOKS,
+    payload: books,
+  });
+};
```

Comment:
- [ ] Kindly note that it is a requirement that you make use of [createasyncthunk](https://redux-toolkit.js.org/api/createAsyncThunk) when making your API calls according to the project requirements. For some context, I will give you an example trying to explain the moving parts as best and as briefly as I can.

```
import { createAsyncThunk } from '@reduxjs/toolkit';

export const getBooks = createAsyncThunk(GET_BOOKS, async () => {
  const response = await fetch(BASE_URL);
  const data = await response.json();

  const newArray = Object.keys(data).map((id) => {
    const obj = data[id][0];
    obj.id = id;
    return obj;
  });

  console.log(newArray);

  return newArray;
});
```

**Explanation**

1. You are first importing the createAsyncThunk into your file to use it. Make sure that you install it first.
2. You are then going to create the getBooks method using the createAsyncThunk. Here the GET_BOOK is simply the action type, nothing fancy
3. Then, because the data returned is of object data structure, I want to convert it into an array and return that array. This step is completely not a must and you could skip it and just return data.

I hope this very brief explanation helps you. Removing and adding a book with follow the same pattern as well.

Reply:
Thanks for example @Nemwel-Boniface , this wasn't easy to implement since i had to refactor some other files but your explanation definitely helped me do so. however i'm glad i did this. learned a few things along way.

![image](https://user-images.githubusercontent.com/66506429/200958302-cb2c836d-2068-465d-acd5-530151638d88.gif)